### PR TITLE
boards: arm: nrf52840-based: Free up flash room for storage/sample apps

### PR DESCRIPTION
**Pull request by mike-scott**
The current flash configurations for all nRF52840's in Zephyr is
VERY constrained when it comes to allowing samples any space for
storage or custom areas.  It only leaves the last 4 pages of flash
for "storage".

The nRF52840 is also capable of using OpenThread which defaults
to using the last 4 pages of flash for storing OpenThread-related
network data.

This means that while using OpenThread under any configuration
designed to use mcuboot partition slots, there is no space left
over for storage of any kind.

Let's reduce the scratch partition size to a reasonable 72k
(down from 120k) and make room for storage, OpenThread and
a bit extra which can be used later when more use cases arise.

Signed-off-by: Michael Scott <mike@foundries.io>

### CODE REVIEW

**Comment by mike-scott** [comment]
For clarity: this patch is needed to use storage AND OpenThread at the same time on nRF52840's.

File: boards/arm/bl654_dvk/bl654_dvk.dts [hunk 1]
```diff
@@ -161,11 +161,21 @@
...
 		};
-		storage_partition: partition@fc000 {
+
+		/*
+		 * The flash starting at 0x000f0000 and ending at
+		 * 0x000fffff is reserved for use by the application.
+		 */
+
+		/*
+		 * If enabled, partition for FCB/NFFS will be created in this
```
Comment (carlescufi) [inline boards/arm/bl654_dvk/bl654_dvk.dts]: I think NVS also uses the storage partition, please add it to the list

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: boards/arm/bl654_dvk/bl654_dvk.dts (hunk 1 above)
Comment (mike-scott) [reply boards/arm/bl654_dvk/bl654_dvk.dts]: Added NVS to the list for all files.

**Comment by mike-scott** [comment]
SHIPPABLE found an overlay file where all of the partition data is copied (not sure why), but it brings up a good point.  I'll do searches on all of these board names and look for .overlay files in our samples/tests.

**Comment by mike-scott** [comment]
Fixed 2 occurrences of pca10056 files with flash partition data.

I'm not sure the DFU test .overlay file does what it needs.  This will *only* affect the Zephyr sample build.  Not the mcuboot build which is based on what is in boards/arm/*.

@carlescufi I think that test is sort of.. broken?  I bet if you were to actually test it on real HW, unless you were manually change the board/arm/nrf52840_pca10056/dts file to match it wouldn't work.

**Comment by nvlsianpu** [comment]
> I think that test is sort of.. broken? I bet if you were to actually test it on real HW, unless you were manually change the board/arm/nrf52840_pca10056/dts file to match it wouldn't work.

No - it is designed to be run on exactly this board. Its .overaly file is specific for the test suite (app must fit into boot partition, ). If someone want run it on other board, then need to create own .overlay file.

Idea of this PR looks good to me - will look closely on Monday.

**Comment by nvlsianpu** [comment]
#### Why scratch partition is so big on nrf52xxx devices?
Each time mcuboot does image swap, it uses scrach partition to swap images. Maximal wear levering of the  swap partition  can be evaluated as images_partition_size/swap_partition_size in relation to wear leveling of image partition. This mean that swap area will be exercised 5.8(3 ) times more than image area using this patch. It is declared that flash on nrf52840 will endure 10000 erase operation. (partially might withstand a magnitude erase more - but this is not guaranteed). This mean that it can withstand at last 1715 DFU operation. For product in a field it is definitely enough. For development purpose might be not to much.

Before this patch ratio was 3.5 - so SoC might withstand at last 2857 DFU operations.

#### How big should be scratch for production firmware?
Depend on requirement - so probably much more smal than for development purpose. For instance: when requirement is to withstand 100 DFU, then we need ratio 100. So especially for nRF52840 this mean starch at size 4.2 KB (so 8 due to page size).

###
My suggestion is to make a .overlay with partition which will be use allowing with examples/samples which are targets of this PR instead of changing scratch size for every sample.

Is this possible technically? What your opinion on that.

**Comment by rlubos** [comment]
@mike-scott OpenThread settings (and it's current oversimplicity) has became an recurring problem recently - I'll look into this and try to propose something smarter.

**Comment by mike-scott** [comment]
@nvlsianpu We can make an overlay in the samples to adjust non-mcuboot related partitions, but currently this is only the last 4 pages of flash (16k).   I know @rlubos brought up that he is working on a better storage method for OpenThread data elements, but some future use case could also require additional custom areas in flash.  For instance, we use 1 page of flash for storing DTLS credential information during development.

I could change this patch to keep the existing size of scratch, and instead:
- take a few (6) pages of flash from each slot
- make the default storage size 0x6000
- keep another 0xA000 for custom additions (in this case, where OpenThread data would land)

This would accomplish the following:
- preserve the development cycle times on the flash chip as was originally intended
- slot size changes from 420k to 396k (to accomplish the above)
- standardizes the nrf52840 storage size to be the same as the nrf52832 (currently it's smaller which would also create more wear)

```
name    start      end        size
------- ---------- ---------- -------------
mcuboot 0x00000000 0x0000BFFF 0xC000
slot0   0x0000C000 0x0006EFFF 0x63000
slot1   0x0006F000 0x000D1FFF 0x63000
scratch 0x000D2000 0x000EFFFF 0x1E000
storage 0x000F0000 0x000F3FFF 0x6000
custom  0x000F6000 0x000FFFFF 0xA000
```

**Comment by lemrey** [comment]
I am unsure why we should change the board DTS to suit a specific use case, in this case OpenThread.
Can this not be done like @nvlsianpu proposed, by using an overlay?

**Comment by mike-scott** [comment]
@lemrey On the flip side of that: IMHO, leaving only 16k for storage is not enough.  It's a powerful chip with quite a lot of use cases which could be combined (trying not to focus too specifically on OpenThread).

Remember: this is the only space that samples can affect easily.. the mcuboot partitions require a changes on both the sample and in mcuboot and are thus much harder to affect.  A simple .overlay in the sample doesn't work to change the mcuboot partitions.

For reference, even the nRF52832 has more storage available:
```
		storage_partition: partition@7a000 {
			label = "storage";
			reg = <0x0007a000 0x00006000>;
		};
```

**Comment by mike-scott** [comment]
@lemrey @nvlsianpu what about 32k (0x8000) for storage (or sample use cases)?

**Comment by nvlsianpu** [comment]
> On the flip side of that: IMHO, leaving only 16k for storage is not enough. It's a powerful chip with quite a lot of use cases which could be combined (trying not to focus too specifically on OpenThread).

> Remember: this is the only space that samples can affect easily.. the mcuboot partitions require a changes on both the sample and in mcuboot and are thus much harder to affect. A simple .overlay in the sample doesn't work to change the mcuboot partitions.

Using  dedicated flash organization is easy, for instance via DTC_OVERLAY_FILE  in command-line.

> What about 32k (0x8000) for storage (or sample use cases)?

I can accept this change. Mainly because nRF52832 has bigger partition.

@lemrey @rlubos Your opinins?

**Comment by lemrey** [comment]
It makes sense that the storage partition on `nrf52840_pca10056` should be at least the same size as the one on `nrf52_pca10040`.

Please be reminded that `nrf52840_pca10059` is different story :)

**Comment by mike-scott** [comment]
Let me update this PR per agreed sizes.

> Please be reminded that nrf52840_pca10059 is different story :)

Yes, this PR purposely avoids pca10059.

**Comment by mike-scott** [comment]
New partition layout looks like this:
```
name    start      end        size
------- ---------- ---------- --------
mcuboot 0x00000000 0x0000BFFF 0xC000
slot0   0x0000C000 0x00072FFF 0x67000
slot1   0x00073000 0x000D9FFF 0x67000
scratch 0x000DA000 0x000F7FFF 0x1E000
storage 0x000F8000 0x000FFFFF 0x8000
```

NOTE: Added nrf52840_papyr to the PR (it was added since the last PR update)

File: tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay [hunk 2]
```diff
@@ -20,19 +20,19 @@
...
-			reg = <0x0000E000 0x000068000>;
+			reg = <0x0000E000 0x000066000>;
 		};
-		slot1_partition: partition@75000 {
+		slot1_partition: partition@73000 {
 			label = "image-1";
-			reg = <0x00076000 0x000068000>;
+			reg = <0x00074000 0x000066000>;
 		};
-		scratch_partition: partition@de000 {
```
Comment (mike-scott) [inline tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay]: NOTE: for an overlay, the scratch and storage partitions don't change and can be removed.  But perhaps this should be in a separate PR.

**Comment by mike-scott** [comment]
SHIPPABLE failing looks like some CI issues regarding WIP commits.  I'll re-run once a few of those are merged.

File: tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay (hunk 2 above)
Comment (nvlsianpu) [reply tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay]: In this test storage partition size is not important - do not modify it then.
Remove this file changes form this PR.

**Comment by nvlsianpu** [comment]
From my side: remove .overlay changes, fix shippable

File: tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay (hunk 2 above)
Comment (mike-scott) [reply tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay]: @nvlsianpu I can't remove all of the overlay changes as we're moving the slot1 and scratch locations.  IE:
`slot1_partition: partition@75000 {` -> `slot1_partition: partition@73000 {`
and
`scratch_partition: partition@de000 {` -> `scratch_partition: partition@da000 {`

For the .overlay to work correctly, it still needs to match on those portions.
I will remove the storage_partition portion of the overlay.

**Comment by mike-scott** [comment]
Let's see if SHIPPABLE issue is now fixed due to merge of fix for philosopher_trace test.

**Comment by mike-scott** [comment]
@nvlsianpu SHIPPABLE failure is complicated and I'm not sure I have an answer.

Looks like NFFS allocates memory structures based on the size of the storage partition.  Currently with 4 pages of flash, the memory usage of the test are:
`SRAM:      258512 B       256 KB     98.61%`

However, when bumped to 8 pages of flash the memory usage overflows like so:
`SRAM:      274896 B       256 KB    104.86%`

**Comment by mike-scott** [comment]
It looks like a mirror of the NFFS file system is kept in RAM?  The added RAM is exactly 4 more pages of flash in size (16k)

**Comment by mike-scott** [comment]
The issue is here where a variable is set to the length of the storage partition.

@nvlsianpu To preserve this test, we should have an overlay for nrf52840_pca10056 to set the storage back to 4 pages of flash (that would be the only setting in the overlay).

**Comment by mike-scott** [comment]
Added an .overlay file to keep the storage size at 16k for the NFFS test which is failing in SHIPPABLE: https://github.com/zephyrproject-rtos/zephyr/pull/13687/files#diff-aecb32e29c0b3f34d9f4a4fd5247ba48R1

Also updated the commit description to include a comment regarding the addition of this .overlay file.

**Comment by mike-scott** [comment]
@nashif SHIPPABLE has some odd "timeout" status.  I don't think this is related to anything this PR did.  Can you take a look?

**Comment by nvlsianpu** [comment]
recheck

File: tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay (hunk 2 above)
Comment (nvlsianpu) [reply tests/subsys/dfu/mcuboot/nrf52840_pca10056.overlay]: right

**Comment by nvlsianpu** [comment]
> It looks like a mirror of the NFFS file system is kept in RAM? The added RAM is exactly 4 more pages of flash in size (16k)

Ram buffer shouldn't be include while flash partition is used - will fix this as it is the bug.

**Comment by nvlsianpu** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/14094

**Comment by mike-scott** [comment]
@nvlsianpu Once that is merged, I will drop the .overlay for that test and then take another look at SHIPPABLE.

**Comment by mike-scott** [comment]
Dropped the overlay added to fix the nffs test.  Let's see how SHIPPABLE does now.

**Comment by mike-scott** [comment]
@carlescufi @nvlsianpu let me know if you need anything else changed with this PR.  SHIPPABLE seems ok now.

**Review by carlescufi — APPROVED** [review]

File: boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts [hunk 3]
```diff
@@ -175,19 +175,29 @@
...
 		};
-		storage_partition: partition@fc000 {
+
+		/*
+		 * The flash starting at 0x000f8000 and ending at
+		 * 0x000fffff is reserved for use by the application.
+		 */
+
+		/*
+		 * If enabled, partition for FCB/NFFS/NVS will be created in
```
Comment (lemrey) [inline boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts]: This is inaccurate, the partition is there regardless and not when only when enabled (partition can't be enabled/disabled).

**Review by lemrey — COMMENTED** [review]
I have left a comment @mike-scott.

File: boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts (hunk 3 above)
Comment (nvlsianpu) [reply boards/arm/nrf52840_pca10056/nrf52840_pca10056.dts]: Thnaks for spot this.
maybe repharse to sth like: storage partition can be used by FCB/NFFS/NVS once one of those is enabled.

**Review by nvlsianpu — APPROVED** [review]

**Review by lemrey — APPROVED** [review]
Thank you.

**Review by ioannisg — APPROVED** [review]
looks ok to me

**Comment by nvlsianpu** [comment]
recheck

**Comment by nvlsianpu** [comment]
recheck

**Review by carlescufi — APPROVED** [review]
